GoogleSignIn/Sources/GIDAppCheck/Implementations/GIDAppCheck.m

```diff
@@ -52,45 +56,64 @@ - (BOOL)isPrepared {
   return [self.userDefaults boolForKey:kGIDAppCheckPreparedKey];
 }
 
-- (void)prepareForAppCheckWithCompletion:(nullable void (^)(FIRAppCheckToken * _Nullable,
-                                                            NSError * _Nullable))completion {
+- (void)prepareForAppCheckWithCompletion:(nullable GIDAppCheckCompletion)completion {
+  if (completion) {
+    @synchronized (self.prepareCompletions) {
+      [self.prepareCompletions addObject:completion];
+    }
+  }
+
+  @synchronized (self) {
+    if (self.preparing) {
+      return;
+    }
+
+    self.preparing = YES;
+  }
+
   dispatch_async(self.workerQueue, ^{
     if ([self isPrepared]) {
       NSError *error = [NSError errorWithDomain:kGIDAppCheckErrorDomain
                                            code:kGIDAppCheckAlreadyPrepared
                                        userInfo:nil];
-      if (completion) {
-        completion(nil, error);
+      @synchronized (self) {
+        for (GIDAppCheckCompletion savedCompletion in self.prepareCompletions) {
+          savedCompletion(nil, error);
```

Comment:
Avoid calling the callback within `@synchornized` because the callback could be heavy and block other code in this class from entering `@synchornized`. You can make an immutable copy of the array in `@synchronized` and invoke the callbacks in the immutable copy outside `@synchronized`.